### DISCUSSION TITLE
mitmproxy 11 QUIC decryption in transparent mode

### DISCUSSION THREAD

**Discussion by emanuele-f** [mitmproxy/mitmproxy#7243] [Q&A] [answered]
Hello, I'm trying running mitmproxy 11 in transparent mode for QUIC decryption, however mitmproxy does not seem to open the UDP port to accept QUIC packets. I've tried the following invocations:

- `mitmdump --mode transparent`
- `mitmdump --mode transparent@8080`
- `mitmdump --mode transparent --set http3=true`

From the https://mitmproxy.org/posts/releases/mitmproxy-11/ post seems like the first invocation should be enough to enable the QUIC support

```
mitmdump --version                
Mitmproxy: 11.0.0
Python:    3.12.7
OpenSSL:   OpenSSL 3.3.2 3 Sep 2024
Platform:  Linux-6.11.3-arch1-1-x86_64-with-glibc2.40
```

**Comment by emanuele-em** [comment]
Maybe the problem is:
> The only major limitation we are aware of at this time is that Chrome [does not trust user-added Certificate Authorities for QUIC](https://issues.chromium.org/issues/40138351#comment15). This means you will either need to provide a publicly trusted certificate (e.g. from Let’s Encrypt), start Chrome with a [command line switch](https://www.chromium.org/quic/playing-with-quic/#generate-certificates), or accept that it falls back to HTTP/2. Alternatively, Firefox doesn’t do such shenanigans. For more HTTP/3 troubleshooting tips, you can check out [#7025](https://github.com/mitmproxy/mitmproxy/issues/7025#issuecomment-2351138170).

**Reply by emanuele-f** [reply]
I'm not yet to that point actually. I've just run mitmproxy and checked netstat for listening UDP sockets, but there are none (neither on port 8080, which is the default listening port, nor 443).

**Comment by mhils** [comment, marked as answer]
Whoops! I think we incorrectly claimed this would be supported in the release blog post. Transparent QUIC interception works in WireGuard and local redirect mode, but transparent mode still misses UDP interception. Sorry about that, I incorrectly added that during copyediting. 😨 

I corrected the blog post for now, will look into how much needs to be done to actually make transparent mode work with that. The hard part - transparent QUIC interception - is certainly done.

**Reply by mhils** [reply]
Some early research notes:

- We basically want to call `recvmsg` with `IP_RECVORIGDSTADDR`. See @meitinger's fantastic original implementation in https://github.com/meitinger/mitmproxy/blob/330243f2abe9fbaa0bd663ea811473f210ab67cf/mitmproxy/net/udp.py#L29
- We now handle UDP in Rust entirely, so we need to do the equivalent there.
- Neither Rust's stdlib nor tokio seem to expose any API for this. There is an ACP for ancilliary data for unix sockets (https://github.com/rust-lang/libs-team/issues/284), but not for UDP sockets.
- I _think_ we need to do something similar to this: 
  https://github.com/tokio-rs/tokio/issues/5417
  https://github.com/pendulum-project/timestamped-socket/blob/369437c965cdbc43c58a96949168f0baeddfbab7/src/socket.rs#L130
  https://github.com/Noratrieb/cluelessh/blob/a03eb384616638df910251631ad7f67cc8f07942/bin/cluelesshd/src/rpc.rs#L644

**Reply by emanuele-f** [reply]
Oh I see, what a pity! I was excited and wanted to integrate this in PCAPdroid, to give users an initial taste of QUIC decryption.

PCAPdroid actually uses the transparent mode only with root decryption, which is still very limited. In non-root mode, PCAPdroid uses the SOCKS5 mode of mitmproxy, and it would be great if mitmproxy supported QUIC decryption in this mode.

Anyway, I take this opportunity to introduce you to an interesting development of the mitmproxy integration in [PCAPdroid-mitm](https://github.com/emanuele-f/PCAPdroid-mitm) is that it's now possible to build and run a full blown version of mitmproxy in Android via Chaquopy (before, I had to patch mitmproxy [removing the mitmproxy_rs module](https://github.com/emanuele-f/mitmproxy/commit/b9549b9186c3e0d2909974335561f9f0707d045f) due to some rust cross compilation issues). This simplifies support for me and also open up new possibilities, like using the wireguard mode of mitmproxy. It would be interesting to allow people to run mitmproxy as a standalone, even without control from pcapdroid. This should be easily doable, maybe some Android use cases will arise in the future on your side 🙂

Also https://github.com/chaquo/chaquopy/pull/1261 should make cross compiling the Cryptography module much easier, which is the main issue when integrating new mitmproxy versions in PCAPdroid. This integration shows the great potential of open source.

Thank you for your great work on this project, it's wonderful to see how much progress has been done in the past months, in particular for QUIC decryption.
